gobblin-runtime/src/main/java/gobblin/runtime/locks/ZookeeperBasedJobLock.java

```diff
@@ -0,0 +1,231 @@
+   * @throws IOException
+   */
+  @Override
+  public boolean tryLock() throws JobLockException {
+    try {
+      return this.lock.acquire(lockAcquireTimeoutMilliseconds, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      throw new JobLockException("Failed to acquire lock " + this.lockPath, e);
+    }
+  }
+
+  /**
+   * Check if the lock is locked.
+   *
+   * @return if the lock is locked
+   * @throws IOException
+   */
+  @Override
+  public boolean isLocked() throws JobLockException {
+    return this.lock.isAcquiredInThisProcess();
+  }
+
+  /**
+   * Closes this stream and releases any system resources associated
+   * with it. If the stream is already closed then invoking this
+   * method has no effect.
+   *
+   * @throws IOException if an I/O error occurs
+   */
+  @Override
+  public void close() throws IOException {
+    try {
+      this.unlock();
+    } catch (JobLockException e) {
+      throw new IOException(e);
+    } finally {
+      lockEventListeners.remove(this.lockPath);
+    }
+  }
+
+  private synchronized static void ensureCuratorFrameworkExists(Properties properties) {
+    if (curatorFramework == null) {
+      CuratorFramework newCuratorFramework = CuratorFrameworkFactory.builder()
+              .connectString(properties.getProperty(CONNECTION_STRING, CONNECTION_STRING_DEFAULT))
+              .connectionTimeoutMs(
+                      getMilliseconds(properties, CONNECTION_TIMEOUT_SECONDS, CONNECTION_TIMEOUT_SECONDS_DEFAULT))
+              .sessionTimeoutMs(
+                      getMilliseconds(properties, SESSION_TIMEOUT_SECONDS, SESSION_TIMEOUT_SECONDS_DEFAULT))
+              .retryPolicy(new ExponentialBackoffRetry(
+                      getMilliseconds(properties, RETRY_BACKOFF_SECONDS, RETRY_BACKOFF_SECONDS_DEFAULT),
+                      getInt(properties, MAX_RETRY_COUNT, MAX_RETRY_COUNT_DEFAULT)))
+              .build();
+
+      newCuratorFramework.getConnectionStateListenable().addListener(new ConnectionStateListener() {
+          @Override
+          public void stateChanged(CuratorFramework curatorFramework, ConnectionState connectionState) {
+            switch (connectionState) {
+              case LOST:
+              case SUSPENDED:
+                log.warn("Lost connection with zookeeper");
+                for (Map.Entry<String, JobLockEventListener> lockEventListener : lockEventListeners.entrySet()) {
+                  log.warn("Informing job %s that lock was lost", lockEventListener.getKey());
+                    lockEventListener.getValue().onLost();
+                }
+                break;
+              case CONNECTED:
+              case RECONNECTED:
+                log.warn("Regained connection with zookeeper");
+                break;
+            }
+          }
+      });
+      newCuratorFramework.start();
+      curatorFramework = newCuratorFramework;
+    }
+    if (locksReaper == null) {
+      ChildReaper newLocksReaper = new ChildReaper(
```

Comment:
Whats this needed for? Also looking at the JavaDocs if we upgrade to Curator 2.9.0 we don't need this anymore: http://curator.apache.org/apidocs/org/apache/curator/framework/recipes/locks/ChildReaper.html